src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -349,11 +349,16 @@ public static void WaitForPendingFinalizers()
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern void _SuppressFinalize(object o);
 
-        public static void SuppressFinalize(object obj)
+        public static unsafe void SuppressFinalize(object obj)
         {
             ArgumentNullException.ThrowIfNull(obj);
 
-            _SuppressFinalize(obj);
+            if (RuntimeHelpers.GetMethodTable(obj)->HasFinalizer)
+            {
+                // SuppressFinalize is a no-op if the object doesn't have a finalizer.
+                // We don't need to call _SuppressFinalize in that case.
```

Comment:
This comment makes it sound like that the HasFinalizer check is just a perf optimization, and it would be ok to call `_SuppressFinalize` unconditionally. I do not think that it is the case. Calling `GCHeapUtilities::GetGCHeap()->SetFinalizationRun` on an object without finalizer would introduce internal inconsistency that's going to lead to assertion failures at minimum.

BTW: `ReRegisterForFinalize` a few lines below has the same HasFinalizer check without any comments.